### PR TITLE
Python typing refinement for dnn_registerLayer/dnn_unregisterLayer functions

### PR DESCRIPTION
**Pull request by VadimLevin**
This patch introduces typings generation for `dnn_registerLayer`/`dnn_unregisterLayer` manually defined in [`cv2/modules/dnn/misc/python/pyopencv_dnn.hpp`](https://github.com/opencv/opencv/blob/4.x/modules/dnn/misc/python/pyopencv_dnn.hpp)

Updates:

- Add `LayerProtocol` to `cv2/dnn/__init__.pyi`:

    ```python
    class LayerProtocol(Protocol):
        def __init__(
            self, params: dict[str, DictValue],
            blobs: typing.Sequence[cv2.typing.MatLike]
        ) -> None: ...

        def getMemoryShapes(
            self, inputs: typing.Sequence[typing.Sequence[int]]
        ) -> typing.Sequence[typing.Sequence[int]]: ...

        def forward(
            self, inputs: typing.Sequence[cv2.typing.MatLike]
        ) -> typing.Sequence[cv2.typing.MatLike]: ...
    ```

- Add `dnn_registerLayer` function to `cv2/__init__.pyi`:

    ```python
    def dnn_registerLayer(layerTypeName: str,
                          layerClass: typing.Type[LayerProtocol]) -> None: ...
    ```

- Add `dnn_unregisterLayer` function to `cv2/__init__.pyi`:

    ```python
    def dnn_unregisterLayer(layerTypeName: str) -> None: ...
    ```
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Review by asmorkalov — APPROVED** [review]
:+1:

File: modules/python/src2/typing_stubs_generation/api_refinement.py [hunk 1]
```diff
@@ -192,6 +195,83 @@ def setMouseCallback(
...
+    """
+    class LayerProtocol(Protocol):
+        def __init__(
+            self, params: dict[str, DictValue],
+            blobs: typing.Sequence[cv2.typing.MatLike]
+        ) -> None: ...
+
+        def getMemoryShapes(
+            self, inputs: typing.Sequence[int]
+        ) -> typing.Sequence[int]: ...
```
Comment (dkurt) [inline modules/python/src2/typing_stubs_generation/api_refinement.py:212]: Expected type is list of lists. Should different typing be used in this way?
https://github.com/opencv/opencv/blob/5c090b9eec0cfe624fb9c3d11e18d9d1b4ff0ead/samples/dnn/edge_detection.py#L35

Comment (VadimLevin) [reply modules/python/src2/typing_stubs_generation/api_refinement.py:212]: Yep its typo, thanks

File: modules/python/src2/typing_stubs_generation/api_refinement.py [hunk 2]
```diff
@@ -192,6 +195,83 @@ def setMouseCallback(
...
+
+    """
+    class LayerProtocol(Protocol):
+        def __init__(
+            self, params: dict[str, DictValue],
+            blobs: typing.Sequence[cv2.typing.MatLike]
+        ) -> None: ...
+
+        def getMemoryShapes(
+            self, inputs: typing.Sequence[int]
```
Comment (dkurt) [inline modules/python/src2/typing_stubs_generation/api_refinement.py:211]: inputs also list of lists

**Pull request merged**